lib/utils/utils.sh

```diff
@@ -1,161 +1,185 @@
-    bpkg-term reset
-  fi
-}
-
-## output error
-bpkg_error () {
-  {
-    bpkg_message "red" "error" "${@}"
-  } >&2
-}
-
-## output warning
-bpkg_warn () {
-  {
-    bpkg_message "yellow" "warn" "${@}"
-  } >&2
-}
-
-## output info
-bpkg_info () {
-  local title="info"
-  if (( "${#}" > 1 )); then
-    title="${1}"
+    export BPKG_PACKAGE_USER="${BPKG_PACKAGE_USER:-"bpkg"}"
+    export BPKG_INDEX=${BPKG_INDEX:-"$HOME/.bpkg/index"}
+
+    bpkg_validate
+
+    return $?
+  }
+
+  ## check parameter consistency
+  bpkg_validate () {
+    if [ ${#BPKG_GIT_REMOTES[@]} -ne ${#BPKG_REMOTES[@]} ]; then
+      bpkg_error "$(printf 'BPKG_GIT_REMOTES[%d] differs in size from BPKG_REMOTES[%d] array' "${#BPKG_GIT_REMOTES[@]}" "${#BPKG_REMOTES[@]}")"
+      return 1
+    fi
+    return 0
+  }
+
+
+  ## format and output message
+  bpkg_message () {
+    bpkg_exec_exist bpkg-term &&
+      bpkg-term color "${1}"
+    shift
+
+    echo -n "    ${1}"
     shift
-  fi
-  bpkg_message "cyan" "${title}" "${@}"
-}
-
-## takes a remote and git-remote and sets the globals:
-##  BPKG_REMOTE: raw remote URI
-##  BPKG_GIT_REMOTE: git remote for cloning
-##  BPKG_AUTH_GIT_REMOTE: git remote with oauth info embedded,
-##  BPKG_OAUTH_TOKEN: token for x-oauth-basic
-##  BPKG_CURL_AUTH_PARAM: auth arguments for raw curl requests
-##  BPKG_REMOTE_INDEX: location of local index for remote
-bpkg_select_remote () {
-  local remote=$1
-  local git_remote=$2
-  export BPKG_REMOTE_HOST=$(echo "$git_remote" | sed 's/.*:\/\///' | sed 's/\/$//' | tr '/' '_')
-  export BPKG_REMOTE_INDEX="$BPKG_INDEX/$BPKG_REMOTE_HOST"
-  # shellcheck disable=SC2034
-  export BPKG_REMOTE_INDEX_FILE="$BPKG_REMOTE_INDEX/index.txt"
-  export BPKG_OAUTH_TOKEN=""
-  export BPKG_CURL_AUTH_PARAM=""
-  export BPKG_GIT_REMOTE=$git_remote
-  export BPKG_AUTH_GIT_REMOTE=$git_remote
-  if [ "${remote:0:10}" == "raw-oauth|" ]; then
-    OLDIFS="${IFS}"
-    IFS="|"
-    # shellcheck disable=SC2206
-    local remote_parts=($remote)
-    IFS="${OLDIFS}"
-    export BPKG_OAUTH_TOKEN=${remote_parts[1]}
+
+    bpkg_exec_exist bpkg-term &&
+      bpkg-term reset
+
+    printf ": "
+
+    bpkg_exec_exist bpkg-term && {
+      bpkg-term reset
+      bpkg-term bright
+    }
+
+    printf "%s\n" "${@}"
+
+    bpkg_exec_exist bpkg-term &&
+      bpkg-term reset
+  }
+
+  ## output error
+  bpkg_error () {
+    {
+      bpkg_message "red" "error" "${@}"
+    } >&2
+  }
+
+  ## output warning
+  bpkg_warn () {
+    {
+      bpkg_message "yellow" "warn" "${@}"
+    } >&2
+  }
+
+  ## output info
+  bpkg_info () {
+    local title="info"
+    if (( "${#}" > 1 )); then
+      title="${1}"
+      shift
+    fi
+    bpkg_message "cyan" "${title}" "${@}"
+  }
+
+
+  ## check if executable exists in path
+  bpkg_exec_exist () {
+    local exec_name="${1}"
+
+    type -f "${exec_name}" > /dev/null 2>&1
+  }
+
+  ## executable exists in path or exit with message
+  bpkg_exec_or_exit () {
+    local exec_name="${1}"
```

Comment:
Might be worth using `"${1?}"` (optionally with a message) to avoid accepting empty values for `$1`.

Reply:
@Potherca: That is a good design choice.  My thought was to maintain the design decision for all the other functions.  I usually use that variable expansion when accepting user input, however since this is an internally used function it seemed like a low likelihood of being improperly used.  If this is a decided change, we should probably implement on other functions also.  Will look at creating a PR with this change across the project.